### ISSUE TITLE
android@9.0.0 and cordova cli 10.0.0  cdvfile://localhost net::ERR_UNKNOWN_URL_SCHEME

### ISSUE THREAD

**Issue by FD-DEV-IT** [apache/cordova-android#1053]
# Bug Report

## Problem

After update to cordova-android@9.0.0 and cordova cli 10.0.0 access to the local storage via cdvfile://localhost is not possible.

### What is expected to happen?

Reading jpg file from local storage to display a image preview

### What does actually happen?

cdvfile://localhost/files/sbsdk-plugin-storage/cee8882b-650f-4107-88c3-5f84f90cbf36.jpg 

net::ERR_UNKNOWN_URL_SCHEME

### Environment, Platform, Device

Cordova Packages:

    cli: 10.0.0
        common: 4.0.2
        create: 3.0.0
        lib: 10.0.0
            common: 4.0.2
            fetch: 3.0.0
            serve: 4.0.0

Project Installed Platforms:

    android: 9.0.0
    browser: null

Project Installed Plugins:

    call-number: 0.0.2
    cordova-android-support-gradle-release: 0.0.2
    cordova-plugin-android-fingerprint-auth: 1.5.0
    cordova-plugin-androidx-adapter: 1.1.1
    cordova-plugin-androidx: 2.0.0
    cordova-plugin-apprate: 1.4.0
    cordova-plugin-battery-status: 2.0.3
    cordova-plugin-camera: 4.1.0
    cordova-plugin-cocoapod-support: 1.3.0
    cordova-plugin-console: 1.1.1-dev
    cordova-plugin-crypto-file: 1.2.1
    cordova-plugin-device: 2.0.3
    cordova-plugin-dialogs: 2.0.2
    cordova-plugin-file-transfer: 1.7.1
    cordova-plugin-file: 6.0.2
    cordova-plugin-firebasex: 10.1.2
    cordova-plugin-globalization: 1.11.0
    cordova-plugin-image-resizer: 1.0.0
    cordova-plugin-inappbrowser: 4.0.0
    cordova-plugin-nativestorage: 2.3.2
    cordova-plugin-network-information: 2.0.2
    cordova-plugin-scanbot-sdk: 4.2.1
    cordova-plugin-splashscreen: 6.0.0
    cordova-plugin-statusbar: 2.4.3
    cordova-plugin-thumbnail: 1.0.0
    cordova-plugin-whitelist: 1.3.4
    cordova.plugins.diagnostic.api-22: 2.3.10-api-22
    phonegap-plugin-barcodescanner: 8.1.0

Environment:

    OS: Microsoft Windows 10 Pro 10.0.18363 (18363) (win32 10.0.18363) x64
    Node: v12.18.3
    npm: 6.14.6

**Comment by FD-DEV-IT** [comment]
Thanks for your answer. The problem is not the path. Apparently the webview blocks the loading of graphics into the html img src

**Comment by FD-DEV-IT** [comment]
The last version (cordova android@8.0.0 /sdk 28) worked without problems. It looks like the cdvfile:// protocol is no longer recognized

**Comment by breautek** [comment]
> Thanks for your answer. The problem is not the path. Apparently the webview blocks the loading of local graphics into the html img src

You're right, you were using a particular part of the feature that I wasn't really familiar with.

I just tested this on `cordova-android@9` targeting API 29. I have an image file inside `cordova.files.dataDirectory + 'files/test.png';`

Which when I use `toInternalURL()`, it gives me `cdvfile://localhost/persistent/test.png` path. This works when used in the DOM, for as long as I update the CSP appropriately.

For me, my [fs root](https://github.com/apache/cordova-plugin-file#cdvfile-protocol) appears to be `persistent`, while your URL seems to have an fs root of `files`

If I change `persistent` to files, I do not receive the `ERR_UNKNOWN_URL_SCHEME`, but I do get a 404 Not found. I did receive `ERR_UNKNOWN_URL_SCHEME` if I use other `localhost/x` paths.

After moving the test image file to the proper place using `adb shell`, I see the `cdvfile://localhost/files/...` works properly for me.

I then created a sub directory and copied the image file, and determine it still works.

I then changed the `AndroidExtraFileSystems` defaults to omit the `files` file-system using:

`<preference name="AndroidExtraFilesystems" value="files-external,documents,sdcard,cache,cache-external,assets,root" />`

This is when I reproduced your issue. So I would check the `AndroidExtraFilesystems` preference and ensure you are including `files`

**Comment by AKempente** [comment]
After trying many of the solutions on this topic, i just want to share how i solved this problem without adding additional configuration meta tags in the header or config.xml in my case i want to take a picture, preview and upload it to server.

the important thing in the following code sample is "url: window.URL.createObjectURL(imgBlob),"
and now i can display the image taken by setting  image DOM src.

I hope this helps someone out there ;)

resolveFile (imageURI) {
      return new Promise((resolve, reject) => {
        window.resolveLocalFileSystemURL(imageURI, (fileEntry) => {
          fileEntry.file(file => {
            const reader = new FileReader()
            reader.onloadend = () => {
              const imgBlob = new Blob([reader.result], {
                type: file.type
              })

              const resData = {
                name: fileEntry.name,
                // url: fileEntry.toInternalURL(), // NO LONGER WORKS
                url: window.URL.createObjectURL(imgBlob),
                blob: imgBlob,
                fileEntry: fileEntry
              }

              window.URL.revokeObjectURL(imgBlob)

              resolve(resData)
            }
            reader.readAsArrayBuffer(file)
          }, (error) => {
            console.log(error)
            reject(error)
          })
        })
      })
    },

**Comment by rutbastoni** [comment]
> After trying many of the solutions on this topic, i just want to share how i solved this problem without adding additional configuration meta tags in the header or config.xml in my case i want to take a picture, preview and upload it to server.
> 
> the important thing in the following code sample is "url: window.URL.createObjectURL(imgBlob)," and now i can display the image taken by setting image DOM src.
> 
> I hope this helps someone out there ;)
> 
> ```
> resolveFile (imageURI) {
>       return new Promise((resolve, reject) => {
>         window.resolveLocalFileSystemURL(imageURI, (fileEntry) => {
>           fileEntry.file(file => {
>             const reader = new FileReader()
>             reader.onloadend = () => {
>               const imgBlob = new Blob([reader.result], {
>                 type: file.type
>               })
> 
>               const resData = {
>                 name: fileEntry.name,
>                 // url: fileEntry.toInternalURL(), // NO LONGER WORKS
>                 url: window.URL.createObjectURL(imgBlob),
>                 blob: imgBlob,
>                 fileEntry: fileEntry
>               }
> 
>               window.URL.revokeObjectURL(imgBlob)
> 
>               resolve(resData)
>             }
>             reader.readAsArrayBuffer(file)
>           }, (error) => {
>             console.log(error)
>             reject(error)
>           })
>         })
>       })
>     }
> ```

Thank you so much, it works also for us

**Comment by breautek** [comment]
Closing.

`cdvfile:` is likely going to become officially deprecated. It has been broken on iOS for some time with the WKWebView, and while not 100% sure what broke it in cordova-android@9, it has also became broken in cordova-android@10 with the WebViewAssetLoader.

We have introduced a newer API to support schemes on android, available with the latest file plugin. Details can be found on the PR: https://github.com/apache/cordova-plugin-file/pull/517
